[FE]스터디룸 그룹장 페이지 - 그룹 정보 및 참가요청 폼 구현

ChanYangYu opened the pull request:
## 관련 이슈

- closes #174

## 셀프 체크리스트

> 최소 1명 이상의 assign을 받아야만 Merge 가능합니다.

- [ ] Warning Message 발생 유/무
- [X] Coding Convention 준수 유/무
- [X] test 통과 여부

## 작업 요약
> 작업 내역 요약
- 그룹장 페이지 전체 기능 구현
## Preview

> 선택 사항

https://user-images.githubusercontent.com/55623688/143178124-cbff725f-6cb6-4035-9b4e-b5fd0e1e9cd9.mp4


## PR 특이 사항

> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점
> 백로그에 없지만 필요하여 구현한 기능 

- 프론트엔드 테스트 코드는 추후에 test 이슈를 하나파서 바로 작성하겠습니다!
- 그룹장 페이지의 백엔드 코드는 현재 다 작성했지만 백엔드 테스트 코드와 같이 작성해서 올리겠습니다!

ChanYangYu committed:
[FE] refactor: datePicker 상태관리 로직 변경


- 외부에서 상태와 상태를 관리하는 함수를 주입하도록 변경하여 재사용이 가능하도록 변경

ChanYangYu committed:
[FE] rename: DatePicker components로 이동


- 기존 GroupCreatePage에 있었던 DatePicker 컴포넌트를 다른곳에서 재사용되기 때문에 components 디렉토리로 이동

ChanYangYu committed:
[FE] feat: 그룹 정보 수정 기능 구현

- 그룹 제목, 소개 : 편집 버튼 클릭시 EditText로 변경되고, 이후 저장버튼 클릭시 정상적으로 텍스트 변경하도록 구현
- 그룹 시작/종료일 : 편집버튼 클릭시 DatePicker가 나오고, 변경 후 저장 버튼 클릭시 정상적으로 텍스트 변경하도록 구현
- issue (#174)

ChanYangYu committed:
[FE] style: 기타 변경사항 적용


- index : groupInfoBox로 네이밍 변경 적용
- ParticipationRequestBox : 박스 높이 변경

ChanYangYu committed:
[FE] feat: 저장 전 값체크 및 알림 메세지 구현


- 그룹 제목 : 1 ~ 20자 텍스트 입력 체크
- 그룹 시작/종료일 : 시작일 종료일이 정상입력되었는지 체크
- 그룹 소개 : 1 ~ 500자 텍스트 입력 체크
- 올바르지 않은 입력시 알림 메세지 출력되도록 구현
- issue (#174)

ChanYangYu committed:
[FE] feat: 그룹장 페이지 Slice 추가


- 그룹장 페이지 전역 상태관리 slice 추가
- issue (#174)

ChanYangYu committed:
[FE] feat: 그룹장 페이지 그룹정보를 GET API 등록


- issue (#174)

ChanYangYu committed:
[FE] feat: 그룹정보 서버 데이터 연결


- 그룹정보 전역상태 저장
- 그룹정보 Date 연결
- issue (#174)

ChanYangYu committed:
[FE] feat: 그룹정보 전역상태 관리


- 화면에 표시되는 그룹정보 전역상태관리
- issue (#174)

ChanYangYu committed:
[FE] feat: 그룹 데이터 PATCH API 등록


- 그룹 데이터 PATCH API 등록
- PATCH에 필요한 ServerData Type 추가
- issue (#174)

ChanYangYu committed:
[FE] feat: 그룹 정보 업데이트 API 연결


- 그룹 이름, 시작/종료일, 그룹 소개 글 업데이트 기능 구현
- issue (#174)

ChanYangYu committed:
[FE] feat: 그룹 참가 요청 GET API 등록


- issue (#174)

ChanYangYu committed:
[FE] feat: 그룹 참가 요청 GET API 연결


- 그룹 참가 요청 리스트 서버 데이터 연결
- issue (#174)

ChanYangYu committed:
[FE] feat: 그룹 참가/수락 거절 API 등록 및 연결


- issue (#174)

ChanYangYu committed:
[FE] feat:  그룹참가 신청 프로필 리다이렉트 구현


- 그룹 참가 신청 리스트에 표시되는 프로필 이미지 클릭시 프로필 페이지로 리다이렉트 되는 기능 구현
- issue (#174)

ChanYangYu committed:
[FE] refactor: 즉시실행함수 적용


- fetchAllData -> 즉시 실행함수로 변경
- issue (#174)

ChanYangYu committed:
[FE] feat: fetch 에러시 홈화면으로 이동 로직 추가


- 권한이 없는 유저 그룹장페이지 접근시 Request 400 에러가 날경우 홈으로 이동시키는 로직 추가
- issue (#174)

Donggggg commented on client/src/api/groupHttpClient.ts:
```diff
@@ -28,6 +33,27 @@ class GroupHttpClient extends HttpClient {
+  public getGroupAdminInfo = (groupId: number): Promise<SimpleGroupInfoResponse> =>
+    this.httpClient.get(`/admin/${groupId}`);
+
+  public patchGroupName = (UpdateGroupNameData: UpdateGroupNameData): Promise<null> =>
+    this.httpClient.patch('/admin/name', UpdateGroupNameData);
+
+  public patchGroupDate = (updateGroupDateData: UpdateGroupDateData): Promise<null> =>
+    this.httpClient.patch('/admin/date', updateGroupDateData);
+
+  public patchGroupIntro = (updateGroupIntroData: UpdateGroupIntroData): Promise<null> =>
+    this.httpClient.patch('/admin/intro', updateGroupIntroData);
+
```
해당 통신 함수들이 데이터를 수정하는 것이라면 `update{데이터명}`이 되면 더 시맨틱할 것 같네용!

ChanYangYu replied:
오 그러네요! 해당 피드백 적용했습니다!

Donggggg commented on client/src/components/DatePicker/index.tsx:
```diff
@@ -5,12 +5,13 @@ import styled from '@emotion/styled';
+  endAt: string,
+  setDate : (startAt: string, endAt: string) => void;
+}
+function AntDatePicker({startAt, endAt, setDate}:Props): JSX.Element {
 
   const checkDate = (current: moment.Moment) => {
     const now = new Date();
```
DatePicker가 2회 이상 사용돼서 컴포넌트로 분리했군요. 좋습니다!!

yuchanleeme commented on client/src/pages/GroupOwnerPage/groupInfoBox/GroupIntro.tsx:
```diff
@@ -0,0 +1,109 @@
+  const { groupId, intro } = useAppSelector(selectGroupAdminData);
+  const dispatch = useAppDispatch();
+  const [isEdit, setIsEdit] = useState<boolean>(false);
+  const textInput = useRef<HTMLTextAreaElement>(null);
```
이 부분은 useState로 관리해도 좋을것 같아요!!

Donggggg replied:
input의 value가 변경될 때 마다 렌더링이 안되는 것을 위해 ref를 사용하셨다면  
input을 ref로 관리하지 않고 useRef로 input의 value값을 관리하면 좋을 것 같아요.

ChanYangYu replied:
useRef 변수로 사용하도록 코드 변경했습니다!

yuchanleeme commented on client/src/pages/GroupOwnerPage/groupInfoBox/GroupTitle.tsx:
```diff
@@ -0,0 +1,107 @@
+  const { groupId, name } = useAppSelector(selectGroupAdminData);
+  const dispatch = useAppDispatch();
+  const [isEdit, setIsEdit] = useState<boolean>(false);
+  const textInput = useRef<HTMLTextAreaElement>(null);
```
여기도 useState로 관리하면 좋을것 같아요!!

ChanYangYu replied:
코드 변경완료했습니다!

ChanYangYu committed:
[FE] rename: 그룹장 페이지 API 이름 변경 


- 그룹장 페이지 API 주소 변경
  - api/group/admin -> api/group-owner
- 그룹장 페이지 Http Client 분리
- issue (#174)

ChanYangYu committed:
[FE] refactor: useRef 변수로 사용 


- useRef를 직접적으로 DomElement를 참조하지 않고 change 이벤트가 발생할 때마다 useRef string 값을업데이트 하는 방식으로 변경
- issue (#174)

yuchanleeme reviewed (approved):
참가요청 폼 구현 수고 많으셨습니다!! 👍

ChanYangYu committed:
[FE] rename: API 함수명 변경

- 피드백 적용
   - patch~  -> update
- issue (#174)
